swap/src/network/tor_transport.rs

```diff
@@ -0,0 +1,225 @@
+    type Output = TcpStream;
+    type Error = io::Error;
+    type Listener = TcpListenStream<Tcp>;
+    type ListenerUpgrade = Ready<Result<Self::Output, Self::Error>>;
+    #[allow(clippy::type_complexity)]
+    type Dial = Pin<Box<dyn Future<Output = Result<Self::Output, Self::Error>> + Send>>;
+
+    fn listen_on(self, addr: Multiaddr) -> Result<Self::Listener, TransportError<Self::Error>> {
+        self.inner.listen_on(addr)
+    }
+
+    // dials via Tor's socks5 proxy if configured and if the provided address is an
+    // onion address. or it falls back to Tcp dialling
+    fn dial(self, addr: Multiaddr) -> Result<Self::Dial, TransportError<Self::Error>> {
+        async fn do_tor_dial(socks_port: u16, dest: String) -> Result<TcpStream, io::Error> {
+            tracing::trace!("Connecting through Tor proxy to address: {}", dest);
+            let stream = connect_to_socks_proxy(dest, socks_port)
+                .await
+                .map_err(|e| io::Error::new(io::ErrorKind::ConnectionRefused, e))?;
+            tracing::trace!("Connection through Tor established");
+            Ok(stream)
+        }
+
+        match to_address_string(addr.clone()) {
+            Some(tor_address_string) => {
+                Ok(Box::pin(do_tor_dial(self.socks_port, tor_address_string)))
+            }
+            _ => {
+                tracing::warn!(
+                    "Address {} could not be formatted. Dialling via clear net",
+                    addr
+                );
+                self.inner.dial(addr)
+            }
+        }
+    }
+
+    fn address_translation(&self, listen: &Multiaddr, observed: &Multiaddr) -> Option<Multiaddr> {
+        self.inner.address_translation(listen, observed)
+    }
+}
+
+/// Tor expects an address format of ADDR:PORT.
+/// This helper function tries to convert the provided multi-address into this
+/// format. None is returned if an unsupported protocol was provided.
+fn to_address_string(multi: Multiaddr) -> Option<String> {
+    let components = multi.iter();
+    for protocol in components {
+        if let Protocol::Onion3(addr) = protocol {
+            return Some(format!(
+                "{}.onion:{}",
+                BASE32.encode(addr.hash()).to_lowercase(),
+                addr.port()
+            ));
+        }
+    }
+
+    // Deal with non-onion addresses
+    let protocols = multi.iter().collect::<Vec<_>>();
+    let address_string = protocols
+        .iter()
+        .filter_map(|protocol| match protocol {
+            Protocol::Ip4(addr) => Some(format!("{}", addr)),
+            Protocol::Ip6(addr) => Some(format!("{}", addr)),
+            Protocol::Dns(addr) => Some(format!("{}", addr)),
+            Protocol::Dns4(addr) => Some(format!("{}", addr)),
+            _ => None,
+        })
+        .collect::<Vec<_>>();
+    if address_string.is_empty() {
+        tracing::warn!(
+            "Could not format address {}. Please consider reporting this issue. ",
```

Comment:
to my understanding this concerns `dnsaddr` not being supported. I don't think we should print that as a warning here, given that we have an additional warning on the caller side. Also I am not sure the `Please consider reporting this issue.` is helpful and the message should include what is the problem more clearly (i.e. `Could not format address {} as onion address (...)`.

Reply:
Unfortunately there are in theory 15 other Protocol types which we did not consider (see a few lines above where we match on `protocol`). 
That's why I just print a warning and hope that someone reports it so that we can improve on the function. 
As of now, I covered all cases I could think of and thought that `Could not format address {}. Please consider reporting this issue` sounds better than `This should never happen` :D